src/components/node-list/index.js

```diff
@@ -1,410 +1,20 @@
-        groupItems.map((item) => item.id),
-        groupItemsDisabled
-      );
-    } else if (isElementType(groupType)) {
-      onToggleTypeDisabled(
-        groupItems.reduce(
-          (state, item) => ({ ...state, [item.id]: !groupItemsDisabled }),
-          {}
-        )
-      );
-    }
-  };
-
-  const handleToggleModularPipelineExpanded = (expanded) => {
-    onToggleModularPipelineExpanded(expanded);
-  };
-
-  const onGroupItemChange = (item, wasChecked) => {
-    // Toggle the group
-    if (isTagType(item.type)) {
-      onToggleTagFilter(item.id, !wasChecked);
-    } else if (isElementType(item.type)) {
-      onToggleTypeDisabled({ [item.id]: wasChecked });
-    }
-
-    // Reset node selection
-    onToggleNodeSelected(null);
-    onToggleNodeActive(null);
-  };
-
-  // Deselect node on Escape key
-  const handleKeyDown = (event) => {
-    if (event.keyCode === 27) {
-      onToggleNodeSelected(null);
-    }
-  };
-
-  // Reset applied filters to default
-  const onResetFilter = () => {
-    onToggleTypeDisabled({ task: false, data: false, parameters: true });
-    onToggleTagFilter(
-      tags.map((item) => item.id),
-      false
-    );
-
-    toUpdateUrlParamsOnResetFilter();
-  };
-
-  // Helper function to check if NodeTypes is modified
-  const hasModifiedNodeTypes = (nodeTypes) => {
-    return nodeTypes.some(
-      (item) => NODE_TYPES[item.id]?.defaultState !== item.disabled
-    );
-  };
-
-  // Updates the reset filter button status based on the node types and tags.
-  useEffect(() => {
-    const isNodeTypeModified = hasModifiedNodeTypes(nodeTypes);
-    const isNodeTagModified = tags.some((tag) => tag.enabled);
-    setIsResetFilterActive(isNodeTypeModified || isNodeTagModified);
-  }, [tags, nodeTypes]);
-
-  useEffect(() => {
-    window.addEventListener('keydown', handleKeyDown);
-    return () => window.removeEventListener('keydown', handleKeyDown);
-  });
-
+const NodeListProvider = ({ faded }) => {
   return (
-    <NodeList
-      faded={faded}
-      items={items}
-      modularPipelinesTree={modularPipelinesTree}
-      modularPipelinesSearchResult={modularPipelinesSearchResult}
-      groups={groups}
-      searchValue={searchValue}
-      onUpdateSearchValue={debounce(updateSearchValue, 250)}
-      onModularPipelineToggleExpanded={handleToggleModularPipelineExpanded}
-      onGroupToggleChanged={onGroupToggleChanged}
-      onToggleGroupCollapsed={onToggleGroupCollapsed}
-      groupCollapsed={groupCollapsed}
-      onToggleFocusMode={onToggleFocusMode}
-      onItemClick={onItemClick}
-      onItemMouseEnter={onItemMouseEnter}
-      onItemMouseLeave={onItemMouseLeave}
-      onToggleHoveredFocusMode={onToggleHoveredFocusMode}
-      onItemChange={onItemChange}
-      focusMode={focusMode}
-      disabledModularPipeline={disabledModularPipeline}
-      onResetFilter={onResetFilter}
-      isResetFilterActive={isResetFilterActive}
-    />
+    <AppContextProvider>
+      <NodeList faded={faded} />
```

Comment:
Using context looks great. One question I have is does NodeList gets rendered everytime there is a change in AppContext whether or not NodeList depends on the change ? I see AppContext has NodeListContextProvider. A bit confused on why can't we have -

```js
 <NodeListContextProvider>
      <NodeList faded={faded} />
```
Thank you

Reply:
Hey, yes, the NodeList component does re-render every time there's a change in AppContext. However, even without context when we just pass everything down as a props through index.js, it will re-render whenever its props change. Good point, though! We could use React.memo to optimise it, but that would be independent of useContext

Reply:
without wrapping it around `AppContext` it will look like this. I thought it looks nicer with just one AppContext. WDYT?

```
 <NodeListContextProvider>
     <FiltersContextProvider>
         <NodeList faded={faded} />
....
```

Reply:
Sorry for the late response. For me AppContext seems more like a generic context, lets say if we want to pass theme of the app or something generic across app. I would rather have individual context or create something like below - 

A group component:

```
<NodePanelContext>
    <NodePanel>
 <NodePanelContext>
 ```
 
 ```
 <NodePanel>
      <PipelineDropDown/>
      <NodeSearch/>
      <NodeList/>
      <FilterList/>
  </NodePanel>
 ```
 
If possible try to have minimum re-renders to children inside NodePanel. For example if someone changes state of FilterList, PipelineDropDown or NodeSearch is not re-rendered . Thank you